### ISSUE TITLE
Cannot play HLS/Fairplay content on Safari with V4 

### ISSUE THREAD

**Issue by sobbe81** [shaka-project/shaka-player#4312]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
Latest, 4.x

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `main`?**
Yes

**Are you using the demo app or your own custom app?**
Custom

**If custom app, can you reproduce the issue using our demo app?**
Cannot test

**What browser and OS are you using?**
Safari on Mac

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**

**What are the manifest and license server URIs?**

**What configuration are you using?  What is the output of `player.getConfiguration()`?**

**What did you do?**

Played a linear channel or a VOD in Safari.

**What did you expect to happen?**
Playback to start

**What actually happened?**

We are having problems playing HLS/Fairplay content with Shaka V4. 

The license request towards the license server seems to work as expected. We originally had issue with that the the defaultInitDataTransform has changed so that skd:// was no longer removed in the data sent in the license request (which V3 does). In addition to successful license requests we also see that manifests and media segments are retrieved from our server but playback never starts. After a long period of time we get an error 3016 saying that media cannot be decoded. We can play the same streams using V3 and we can also play unencrypted versions of the streams using V4.

Comparing the logs between V3 and V4 we see differences when it comes to implementation of MediaKeys. If I understand correctly, V4 is using the EME for handling the media keys where as V3 is using a poly-filled version?
 
We are out of ideas on how to troubleshoot this further. Do you have any suggestions? 

Is it possible to configure the implementation of MediaKeys in V4 to at least test to see if that is the root cause?

**Comment by hiren3897** [comment]
Duplicate #4211

I had a same issue

**Comment by avelad** [comment]
It will be resolved with https://github.com/shaka-project/shaka-player/pull/4309

**Comment by fredrik-telia** [comment]
We're currently on shaka 3.1.0 in our project, trying to upgrade to 4.2.0 (issues with manifests that go from dynamic->static have prevented us from upgrading so far)

It is mostly working, except in the case of HLS/Safari:

When using native playback in Safari (`useNativeHlsOnSafari: true`) we get intermittent issues, with `"Media failed to decode"` before we even see a frame of the video, this happens like 95% of the time.. the other 5% of the time video plays. Debugging still ongoing to find out more as to why this happens.

When using Shaka for playback (`useNativeHlsOnSafari: false`) everything works fine, EXCEPT when an ongoing recording changes from a dynamic hls manifest to a static one. This issue is also present on Chrome when playing HLS.
Here's an example scenario:

Program A starts at 10:00 and ends at 11:00
I start the program from the beginning at wall clock time 10:40, the live point in the progress bar is now at 40 minutes.
I'm watching the program, and the live point (`seekRange.end`) moves to the right.

At 11:00 the recording ends, and the manifest goes static, the playhead is at 20 minutes. The live point will now "snap back" to what it was when I started playing, so back to 40 minutes.

When the playhead reaches this point, the buffer will drop to 0 and the stream will stall.

As a hacky workaround in trying to understand this problem, I've changed the `Math.min...` in `getSegmentAvailabilityEnd()` in the file `presentation_timeline.js` to be a `Math.max...` and that will allow me to play the content all the way through.

Maybe this means that `this.maxSegmentEndTime_` is not updated correctly when HLS manifests go from dynamic to static?
